x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/mapper/DenseVectorLSHFieldMapper.java

```diff
@@ -0,0 +1,393 @@
+        super(simpleName, fieldType, defaultFieldType, indexSettings, multiFields, copyTo);
+    }
+
+    @Override
+    protected DenseVectorLSHFieldMapper clone() {
+        return (DenseVectorLSHFieldMapper) super.clone();
+    }
+
+    @Override
+    public DenseVectorLSHFieldType fieldType() {
+        return (DenseVectorLSHFieldType) super.fieldType();
+    }
+
+    @Override
+    public void parse(ParseContext context) throws IOException {
+        // TODO: implement usage in multi-fields
+        if (context.externalValueSet()) {
+            throw new IllegalArgumentException("Field [" + name() + "] of type [" + typeName() + "] can't be used in multi-fields");
+        }
+
+        int dims = fieldType().lshModel.dims(); // number of vector dimensions
+        float[] vector = new float[dims];
+        int dim = 0;
+        for (Token token = context.parser().nextToken(); token != Token.END_ARRAY; token = context.parser().nextToken()) {
+            if (dim >= dims) {
+                throw new IllegalArgumentException("Field [" + name() + "] of type [" + typeName() + "] of doc [" + context.doc() +
+                    "] has number of dimensions [" + dim + "] greater than defined in the mapping: [" +  dims +"]");
+            }
+            ensureExpectedToken(Token.VALUE_NUMBER, token, context.parser()::getTokenLocation);
+            vector[dim] = context.parser().floatValue(true);
+            dim++;
+        }
+        if (dim != dims) {
+            throw new IllegalArgumentException("Field [" + name() + "] of type [" + typeName() + "] of doc [" + context.doc() +
+                "] has number of dimensions [" + dim + "] less than defined in the mapping: [" +  dims +"]");
+        }
+
+        // hash vector to get L number of hashes corresponding to L hash tables
+        BytesRef[] hashes = fieldType().lshModel().hash(vector);
+        // index each hash as an untokenized term
+        String fieldNameLSH = fieldType().name();
+        for (int i = 0; i < hashes.length; i++) {
+            Field indexedField = new Field(fieldNameLSH, hashes[i], fieldType());
+            context.doc().add(indexedField);
+        }
+    }
+
+    @Override
+    protected void parseCreateField(ParseContext context, List<IndexableField> fields) {
+        throw new AssertionError("parse is implemented directly");
+    }
+
+    @Override
+    protected String contentType() {
+        return CONTENT_TYPE;
+    }
+
+    @Override
+    protected void doXContentBody(XContentBuilder builder, boolean includeDefaults, Params params) throws IOException {
+        super.doXContentBody(builder, includeDefaults, params);
+        builder.field("dims", fieldType().dims());
+        builder.field("l", fieldType().lshModel.l());
+        builder.field("k", fieldType().lshModel.k());
+        builder.field("w", fieldType().lshModel.w());
+        builder.field("a", fieldType().lshModel.a());
```

Comment:
You are absolutely right about this.
For `audio` data, for example:
8 (l) * 20 (k) * 192 (d) * 4(bytes per float) = 122880 bytes = 120Mb . 
This could be 5 times bigger for `1024` dims.

I just don't know how to store the model otherwise.  Storing in a file creates problems how to synchronize this file across nodes/replicas.  This could be a good topic to discuss on our ann meeting.